File: ios/bookz/bookz/home/GenreType.swift
//
//  GenreType.swift
//  bookz
//
//  Created by Aldana, Sal on 1/18/23.
//

import Foundation
import SwiftUI
import DynamicColor

enum GenreType: Int, CaseIterable {
    case adventure, fantasy,
         scifi, romance
    
    var displayLabel: String {
        switch self {
        case .adventure:
            return "Adventure"
        case .fantasy:
            return "Fantasy"
        case .scifi:
            return "Sci-Fi"
        case .romance:
            return "Romance"
        }
    }
    
    var searchString: String {
        var option: String
        switch self {
        case .adventure:
            option = "adventure"
        case .fantasy:
            option = "fantasy"
        case .scifi:
            option = "science%20fiction"
        case .romance:
            return "romance"
        }
        return "+subject:\(option)"
    }
    
    var primaryColor: Color {
        switch self {
        case .adventure:
            return DynamicColor(hexString: "68e1fd").swiftColor
        case .fantasy:
            return DynamicColor(hexString: "fca868").swiftColor
        case .scifi:
            return DynamicColor(hexString: "68fc78").swiftColor
        case .romance:
            return DynamicColor(hexString: "fc68d7").swiftColor
        }
    }
    
    var gradient: LinearGradient {
        switch self {
        case .adventure:
            return self.primaryColor.topDownGradient
        case .fantasy:
            return self.primaryColor.leftRightGradient
        case .scifi:
            return self.primaryColor.bottomUpGradient
        case .romance:
            return self.primaryColor.rightLeftGradient
        }
    }
    
    var imageName: String {
        switch self {
        case .adventure:
            return "Adventure"
        case .fantasy:
            return "Fortnite"
        case .scifi:
            return "Alien"
        case .romance:
            return "Couple"
        }
    }
}

File: android/app/src/main/java/net/level3studios/bookz/models/GenreType.kt
package net.level3studios.bookz.models

import androidx.compose.material.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import net.level3studios.bookz.R
import net.level3studios.bookz.ui.theme.*

enum class GenreType(val id: Int) {
    ADVENTURE(0),
    FANTASY(1),
    SCIFI(2),
    ROMANCE(3);

    companion object  {
        fun getType(id: Int): GenreType {
            return when (id) {
                ADVENTURE.id -> ADVENTURE
                FANTASY.id -> FANTASY
                SCIFI.id -> SCIFI
                ROMANCE.id -> ROMANCE
                else -> ADVENTURE
            }
        }
        @Composable
        fun GenreType.getPrimaryColor(): Color {
            return when (this) {
                ADVENTURE -> MaterialTheme.colors.customBlue
                FANTASY -> MaterialTheme.colors.customOrange
                SCIFI -> MaterialTheme.colors.customGreen
                ROMANCE -> MaterialTheme.colors.customPink
            }
        }
        @Composable
        fun GenreType.getPrimaryContainer(): Color {
            return when (this) {
                ADVENTURE -> MaterialTheme.colors.customBlueContainer
                FANTASY -> MaterialTheme.colors.customOrangeContainer
                SCIFI -> MaterialTheme.colors.customGreenContainer
                ROMANCE -> MaterialTheme.colors.customPinkContainer
            }
        }

        fun GenreType.getImageId(): Int {
            return when (this) {
                ADVENTURE -> R.drawable.adventure
                FANTASY -> R.drawable.fortnite
                SCIFI -> R.drawable.alien
                ROMANCE -> R.drawable.couple
            }
        }
        fun GenreType.getDisplayLabel(): String {
            return when (this) {
                ADVENTURE -> "Adventure"
                FANTASY -> "Fantasy"
                SCIFI -> "Sci-Fi"
                ROMANCE -> "Romance"
            }
        }
        fun GenreType.getSearchString(): String {
            val option: String = when (this) {
                ADVENTURE -> "adventure"
                FANTASY -> "fantasy"
                SCIFI -> "science%20fiction"
                ROMANCE -> "romance"
            }
            return "+subject:$option"
        }
    }
}